### PR TITLE
Refactor rewards engine URL requests

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/35246

* Moves URL loading and logging helpers out of `RewardsEngineImpl` into a separate class.
* Removes ability to supply `std::function` to the URL loading helper.

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_rewards/core/endpoint/gemini/post_account/post_account_gemini.cc [hunk 1]
```diff
@@ -85,15 +86,15 @@ void PostAccount::Request(const std::string& token,
   request->headers = RequestAuthorization(token);
   request->method = mojom::UrlMethod::POST;
 
-  engine_->LoadURL(std::move(request),
-                   base::BindOnce(&PostAccount::OnRequest,
-                                  base::Unretained(this), std::move(callback)));
+  engine_->Get<URLLoader>().Load(
+      std::move(request), URLLoader::LogLevel::kDetailed,
+      base::BindOnce(&PostAccount::OnRequest, base::Unretained(this),
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply components/brave_rewards/core/endpoint/gemini/post_account/post_account_gemini.cc:91]: This is due to converting existing `std::function` code to use `base::Callback`. `base::Unretained` will not be used for new code.

**Review by emerick — APPROVED** [review]

File: components/brave_rewards/core/common/url_loader.cc [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      std::move(request),
+      base::BindOnce(&URLLoader::OnResponse, weak_factory_.GetWeakPtr(),
+                     log_level, std::move(callback)));
+}
+
+bool URLLoader::ShouldLogRequestHeader(const std::string& header) {
+  constexpr std::array allowed_headers{"digest", "signature", "accept",
+                                       "content-type"};
+
+  for (auto* name : allowed_headers) {
```
Comment (szilardszaloki) [inline components/brave_rewards/core/common/url_loader.cc:57]: You can also specify these in the loop itself:
```cpp
for (auto* name : {"digest", "signature", "accept", "content-type"}) {
  // ...
}
```

Comment (szilardszaloki) [reply components/brave_rewards/core/common/url_loader.cc:57]: Or use `std/base::ranges::any_of<>()`.

Comment (zenparsing) [reply components/brave_rewards/core/common/url_loader.cc:57]: Cool - I think probably the existing variant is the most self-documenting, though.

Comment (szilardszaloki) [reply components/brave_rewards/core/common/url_loader.cc:57]: Yeah, that's fine, too. 👍🏻

File: components/brave_rewards/core/endpoint/bitflyer/post_oauth/post_oauth_bitflyer.cc [hunk 3]
```diff
@@ -110,17 +111,16 @@ void PostOauth::Request(const std::string& external_account_id,
   request->headers = RequestAuthorization();
   request->content_type = "application/json";
   request->method = mojom::UrlMethod::POST;
-  request->skip_log = true;
 
-  engine_->LoadURL(std::move(request),
-                   base::BindOnce(&PostOauth::OnRequest, base::Unretained(this),
-                                  std::move(callback)));
+  engine_->Get<URLLoader>().Load(
+      std::move(request), URLLoader::LogLevel::kBasic,
```
Comment (szilardszaloki) [inline components/brave_rewards/core/endpoint/bitflyer/post_oauth/post_oauth_bitflyer.cc:116]: Should this be `kNone` instead?

Comment (zenparsing) [reply components/brave_rewards/core/endpoint/bitflyer/post_oauth/post_oauth_bitflyer.cc:116]: Perhaps. The current implementation would effectively skip logging of both requests and responses, but it seems like it's useful to still log the basic info (URL/method). Open to not logging anything here to match current behavior though. Probably the safest approach.

Comment (zenparsing) [reply components/brave_rewards/core/endpoint/bitflyer/post_oauth/post_oauth_bitflyer.cc]: I've updated this and similar calls to use `kNone`, to err on the side of caution. Great catch!

Comment (szilardszaloki) [reply components/brave_rewards/core/endpoint/bitflyer/post_oauth/post_oauth_bitflyer.cc]: Yeah, I saw `skip_log` being mapped to both `kNone`, and `kBasic`, so I thought it's better to ask. As long as we don't log sensitive stuff, I'm good with either, but might be safest to stick with what we have, as you said. 🙂

**Review by szilardszaloki — APPROVED** [review]

**Pull request merged**